### DISCUSSION TITLE
Filtering or patching assemblies

### DISCUSSION THREAD

**Discussion by deliciouslytyped** [jbevain/cecil#866] [Q&A]
Hey all,

I'm trying to extract some code that has non-system-dependent business logic from Xamarin binaries to run them on Linux.
What would be the best way to go about this?

I expect that I could just run the code from the original assembly directly, but the mono loader (understandably) complains if I try to remove the Android related assemblies, because the business logic assembly links against them. This seems to mean for example references in the AssemblyRef tables, and the AssemblyRef table data is used by the TypeRef table data and whatnot.

It would be fine with me if I could just remove all the information I want to avoid (the things referring to the android platform specific libraries) from the metadata tables, but if I understand correctly, I would still have to update the tokens referring to the table entries in the CIL asm.

Also some of the code uses System.IO for networking (IIRC), but I think I should just be able to swap in another platform specific .net standard library, if I can manage to somehow re-engineer the binary I have.
I guess I would also need to link in the new platform specific assemblies I want? - by updating the tables and the CIL asm.

I have several ideas, but without knowing how Cecil is put together, I'd just end up knee deep in the X-Y problem.

I wanted to avoid re-generating the entire binary to minimize the difference between my patch and the original library, but given that Cecil works by re-generating the entire binary, that doesn't really seem to be an option. So maybe I don't need to worry so much about low level details and just filter out/replace the things I don't want at a higher level?

xref: https://github.com/jbevain/cecil/issues/865#issuecomment-1169328318

**Comment by deliciouslytyped** [comment]
"Basically", I'd like to swap a specific set of assembly dependencies for a different set, and ignore any remaining dependencies.
I only want to run a subset of the code, and hopefully that doesn't need the things that are missing.

**Comment by ltrzesniewski** [comment]
> It would be fine with me if I could just remove all the information I want to avoid (the things referring to the android platform specific libraries) from the metadata tables, but if I understand correctly, I would still have to update the tokens referring to the table entries in the CIL asm.

If something's in the metadata tables it (usually) means it's used in the CIL stream. If you simply remove something from the metadata tables, the metadata tokens referenced by the IL stream won't be valid anymore.

> "Basically", I'd like to swap a specific set of assembly dependencies for a different set, and ignore any remaining dependencies.

If you want to avoid having to update the IL, I suppose the "easiest" approach would be to change the `AssemblyRef` table to substitute the reference you want to remove with your own shim assembly which would expose the same API.

Cecil doesn't expose the raw metadata tables directly. Instead, it provides a higher-level view of the assembly as an object graph. You work with object references instead of metadata tokens.

I think you should be able to modify the `AssemblyNameReference` object in `ModuleDefinition.AssemblyReferences` to point to your own assembly. Be sure to *modify* the existing object instead of removing it and inserting your own, otherwise you'd have to update every reference to the old instance (the assembly is exposed as a *giant* object graph you'd rather not iterate over).

Side note: while `System.Reflection.Metadata` gives you raw access to the metadata tables, I don't think you'd be able to patch them without rewriting the whole assembly anyway, since you need to allocate a string blob for your new assembly name. dnSpy also lets you edit the metadata tables and the existing `#Strings` heap contents, but won't let you insert a new string to the heap for the same reason.

**Comment by deliciouslytyped** [comment]
Thanks, this is helpful.
 I'm not sure when I'll have a chance to work on this again, but I'll definitely reply here when I make some progress.

**Reply by deliciouslytyped** [reply]
@ltrzesniewski 
This is where I meant to ping you earlier instead of https://github.com/jbevain/cecil/issues/865#issuecomment-1213189514 , sorry.

**Reply by deliciouslytyped** [reply]
You seem to know your way around Cecil and CLI so I thought I might ask in more detail about my project.

I also just figured out this discussion thing is supposed to be like the stackoverflow QA format and not necessarily a standard linear top down thread. Oh well.
